### ISSUE TITLE
Create development environment for API

### ISSUE THREAD

**Issue by dereklieu** [IFRCGo/go-frontend#479]
It's clear to me that we need to create a development environment for the API. This is for a couple of reasons:

1. Locally, it's hard to test certain aspects of the API, namely the NGINX proxy and production webserver setup, because we use Python's development server locally. This makes it hard to debug, for example, upload size limits #206.

1. It can be useful to have an environment to run certain tasks on, such as the scrapers, that are closer to the staging and production environments.

1. Without a development environment, such debugging gets relegated to the staging environment.

Lately, development using the staging environment has resulted in the staging site losing some core functionality, degrading our ability to confirm new features and bug fixes. We cannot use the staging environment for this purpose, especially not for one-off experiments. Mixing staging and development means we will eventually get to a point where we cannot trust the staging environment to show us what a feature or bug fix would look like on the production environment, essentially nullifying it's usefulness.

I propose a separate Azure VM instance hosting both a Postgres DB and connected API. The new development API will use existing storage containers and keys, meaning anyone who has access to staging/production will also have access to dev. The new environment will not be hooked up to any front-end by default, but a local front-end instance can connect to the development API by setting the proper URL.

By using existing storage containers and cohosting the DB and API, the cost of the dev environment will equal one Azure VM, as low as $15/mo.

Updating the API instance to connect to a local DB while setting up NGINX webservers I estimate to be about a day (8 hours) of my time. What this would achieve is the ability to easily deploy a self-contained development infrastructure where someone could test features closer to the target environment, without contaminating the current staging environment.

@LukeCaley thoughts?

**Comment by LukeCaley** [comment]
Thanks @dereklieu.  It seems clear that creation of a development environment is necessary to include in the scope of the next phase, planned for Q1 2019.  However, I'm not sure if this is absolutely necessary for the small bug fixes and maintenance issues we are addressing right now, especially given the cost for the remaining time on your contract.

**Comment by nanometrenat** [comment]
@batpad please review and advise necessity/priority of this.

**Comment by batpad** [comment]
cc @thenav56 I know you guys have been doing some work standing up separate development instances for 3w. Would you be able to advise a path forward here? This is still something that would be nice - we now have the `surge` URLs that covers this for the frontend, but on the backend we are restricted to staging and production.

**Comment by thenav56** [comment]
@batpad
There are some questions I have before we can move forward.
This issue's original idea is to add a testing server which should be a small scale replica of the production server where one can push a testing instance of the **go-api**.
This is nice to have but looking at the current development of the go, I feel like this will not cover all the development process properly for eg: we have **multiple teams** working on separate features/fixes, so whose instance should be available in the testing server? and for how long?


My suggestion will be to just create a simple VM instance where teams have their separate
- instances 
- databases (same database instance but different database names)
- domains (to connect with the surge deploys)

This way the VM instance will be simple, 1 proxy server ([caddy](https://caddyserver.com/) will be fine to use here), multiple **go-api** docker instances, and a single (or multiple) databases (using docker is fine). Also, no need to connect to Azura services like storage. Something like this
```yaml
version: '3.2'

services:
    # IFRC GO DB
    go-db:
      image: postgres:9.6
      environment:
        POSTGRES_PASSWORD: test
        POSTGRES_USER: test
        POSTGRES_DB: test
      volumes:
        - ./go-db/.tmp/pg:/var/lib/postgresql/data

    # IFRC GO TC Instance
    go-api-tc:
      image: ifrcgo/go-api-tc:latest
      build: go-api-tc/
      environment:
        DJANGO_DB_NAME: test
        DJANGO_SECRET_KEY: test
        DJANGO_DB_HOST: go-db
        DJANGO_DB_USER: test
        DJANGO_DB_PASS: test
        FRONTEND_URL: tc-ifrc-go.surge.sh
        # FDRS_CREDENTIAL:
        # HPC_CREDENTIAL:
      volumes:
        - ./go-api-tc/:/home/ifrc/go-api
      command: python manage.py runserver 0.0.0.0:8000
      restart: unless-stopped
      depends_on:
        - go-db

    caddy: # Caddy Server
      image: abiosoft/caddy:1.0.3
      volumes:
        # Volumes between caddy and other apps
        - ./go-api-tc/static/:/go-api/staic/tc
        - ./go-api-tc/medial/:/go-api/media/tc
        # Caddy config volumes
        - ./caddy/Caddyfile:/etc/Caddyfile
        - ./caddy/.caddy:/root/.caddy
      ports:
        - '80:80'
        - '443:443'
      depends_on:
        - go-api-tc
```

This way they can maintain/update their instance as required without interfering with (or wait for) other's work.
Still, this will miss some features like elastic search or any external services since setting those up for each instance will be time/cost consuming.

Let me know what you think.
